{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1f984820",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\marce\\AppData\\Local\\Temp\\ipykernel_12076\\1323507865.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ab_ag[\"CDR_H1\"] = np.nan\n",
      "C:\\Users\\marce\\AppData\\Local\\Temp\\ipykernel_12076\\1323507865.py:16: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ab_ag[\"CDR_H2\"] = np.nan\n",
      "C:\\Users\\marce\\AppData\\Local\\Temp\\ipykernel_12076\\1323507865.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ab_ag[\"CDR_H3\"] = np.nan\n",
      "C:\\Users\\marce\\AppData\\Local\\Temp\\ipykernel_12076\\1323507865.py:18: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ab_ag[\"CDR_L1\"] = np.nan\n",
      "C:\\Users\\marce\\AppData\\Local\\Temp\\ipykernel_12076\\1323507865.py:19: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ab_ag[\"CDR_L2\"] = np.nan\n",
      "C:\\Users\\marce\\AppData\\Local\\Temp\\ipykernel_12076\\1323507865.py:20: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ab_ag[\"CDR_L3\"] = np.nan\n",
      "C:\\Users\\marce\\AppData\\Local\\Temp\\ipykernel_12076\\1323507865.py:22: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ab_ag.dropna(subset=[\"antigen_chain\"], inplace=True)\n"
     ]
    }
   ],
   "source": [
    "# Notebook vorbereiten\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import os\n",
    "import shutil \n",
    "ab_ag_raw = pd.read_csv('data/ab_ag.tsv', sep='\\t')\n",
    "columns = pd.read_csv('data/columns.tsv', sep='\\t')\n",
    "uniprot_data = pd.read_csv('data/uniprot_data.tsv', sep='\\t')\n",
    "\n",
    "# ab_ag aus ab_ag_raw aufbereiten\n",
    "ab_ag = ab_ag_raw[[\"pdb\", \"Hchain\", \"Lchain\", \"model\", \"antigen_chain\", \"antigen_name\", \"antigen_species\"]]\n",
    "# Füge Spalten für die CDR-Sequenzen in ab_ag hinzu\n",
    "ab_ag[\"CDR_H1\"] = np.nan\n",
    "ab_ag[\"CDR_H2\"] = np.nan\n",
    "ab_ag[\"CDR_H3\"] = np.nan\n",
    "ab_ag[\"CDR_L1\"] = np.nan\n",
    "ab_ag[\"CDR_L2\"] = np.nan\n",
    "ab_ag[\"CDR_L3\"] = np.nan\n",
    "# Entferne Einträge ohne Antigennamen\n",
    "ab_ag.dropna(subset=[\"antigen_chain\"], inplace=True)\n",
    "\n",
    "# Ich hab aus Versehen das Code-Feld gelöscht! Code von Anna für doppelte pdb Einträge zurückholen!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f56aec97",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Could not download Chothia-annotated sequence for 9ds2: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 9ds2: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 9ds2: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8uzp: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8uzp: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8veb: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8veb: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8veb: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8ved: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8ved: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8ved: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8vee: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8vee: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8vee: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8vef: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8vef: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8vef: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8vpf: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8vpf: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8vpf: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 9dpc: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 9dru: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 9ds1: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 9jhj: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 9l6c: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 9l6c: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 9l6c: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 9mer: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 9mev: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8rmx: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8rmx: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8rmy: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8rmy: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8u1p: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8u1p: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8yk4: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8ykg: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8ykg: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8ykh: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 8ykt: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 9azr: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 9azt: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 9azv: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 9bjg: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 9bjh: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 9bjh: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 9cci: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 9ccj: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 9j66: HTTP Error 404: Not Found\n",
      "Could not download Chothia-annotated sequence for 9knw: HTTP Error 404: Not Found\n"
     ]
    }
   ],
   "source": [
    "# Überarbeiteter Code zum Herunterladen der Antikörpersequenzen. Direkt in ab_ag_raw speichern\n",
    "# Funktionen aus dem sabdab_downloader importieren\n",
    "from common.sabdab_downloader import getannotation\n",
    "\n",
    "# Definiere CDR-Regionen als slices für die H- und L-Ketten (Literatur überprüfen, ob die Indizes korrekt sind!)\n",
    "cdr_regions = {\n",
    "        \"H1\": slice(26, 32),\n",
    "        \"H2\": slice(52, 56),\n",
    "        \"H3\": slice(95, 102),\n",
    "        \"L1\": slice(24, 34),\n",
    "        \"L2\": slice(50, 56),\n",
    "        \"L3\": slice(89, 97)\n",
    "}\n",
    "\n",
    "# Erstelle eine Liste mit allen PDB-Einträgen aus ab_ag (enthält Duplikate)\n",
    "pdb_entries = ab_ag[\"pdb\"].tolist()\n",
    "\n",
    "# Lade die Chothia-Daten und Chothia-nummerierten Sequenzen für jeden PDB-Eintrag herunter und speichere sie im Ordner des entsprechenden PDB-Eintrags\n",
    "for x in range(0,50): # Testlauf, nur die ersten 5 PDB-Einträge\n",
    "    # Zeilenweise Einträge für den aktuellen PDB-Eintrag extrahieren (für die Chothia-annotierten Sequenzen)\n",
    "    fab_list = ab_ag.iloc[x,:].values.tolist()\n",
    "    pdb = fab_list[0] # PDB-ID aus der Liste extrahieren\n",
    "    # Erstelle den Ordner für die Chothia-annotierte Sequenz des PDB-Eintrags, falls er nicht existiert\n",
    "    if not os.path.exists(f\"data/sabdab_download/{pdb}\"):\n",
    "        os.makedirs(f\"data/sabdab_download/{pdb}\")\n",
    "    # Lade die Chothia-nummerierten Sequenzen herunter oder gib eine Fehlermeldung für entsprechende PDB-Einträge aus\n",
    "    try:\n",
    "        getannotation(pdb, fab_list, f\"data/sabdab_download/{pdb}\")\n",
    "        # Extrahiere die CDR-Sequenzen aus den heruntergeladenen Chothia-annotierten Sequenzen und speichere sie in ab_ag\n",
    "        # Speichere jede .ann-Datei in einm DataFrame\n",
    "        VH_chothia_seq = pd.read_csv(f\"data/sabdab_download/{pdb}/{pdb}_{fab_list[1]}_VH.ann\", sep='\\t', header=None, names=['Chothia_Position', 'Aminosäure'])\n",
    "        VL_chothia_seq = pd.read_csv(f\"data/sabdab_download/{pdb}/{pdb}_{fab_list[2]}_VL.ann\", sep='\\t', header=None, names=['Chothia_Position', 'Aminosäure'])\n",
    "        # Extrahiere die CDR-Sequenzen aus den Chothia-annotierten Sequenzen\n",
    "        H1_seq = VH_chothia_seq.loc[cdr_regions[\"H1\"], \"Aminosäure\"].tolist()\n",
    "        ab_ag.loc[ab_ag[\"pdb\"] == pdb, \"CDR_H1\"] = H1_seq\n",
    "        H2_seq = VH_chothia_seq.loc[cdr_regions[\"H2\"], \"Aminosäure\"].tolist()\n",
    "        ab_ag.loc[ab_ag[\"pdb\"] == pdb, \"CDR_H2\"] = H1_seq\n",
    "        H3_seq = VH_chothia_seq.loc[cdr_regions[\"H3\"], \"Aminosäure\"].tolist()\n",
    "        ab_ag.loc[ab_ag[\"pdb\"] == pdb, \"CDR_H3\"] = H1_seq\n",
    "        L1_seq = VL_chothia_seq.loc[cdr_regions[\"L1\"], \"Aminosäure\"].tolist()\n",
    "        ab_ag.loc[ab_ag[\"pdb\"] == pdb, \"CDR_L1\"] = H1_seq\n",
    "        L2_seq = VL_chothia_seq.loc[cdr_regions[\"L2\"], \"Aminosäure\"].tolist()\n",
    "        ab_ag.loc[ab_ag[\"pdb\"] == pdb, \"CDR_L2\"] = H1_seq\n",
    "        L3_seq = VL_chothia_seq.loc[cdr_regions[\"L3\"], \"Aminosäure\"].tolist()\n",
    "        ab_ag.loc[ab_ag[\"pdb\"] == pdb, \"CDR_L3\"] = H1_seq\n",
    "    except Exception as e:\n",
    "        print(f\"Could not download Chothia-annotated sequence for {pdb}: {e}\")\n",
    "    shutil.rmtree(f\"data/sabdab_download/{pdb}\")  # Lösche den Ordner, nachdem die Sequenzen extrahiert wurden"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5adb98bc",
   "metadata": {},
   "outputs": [
    {
     "ename": "HTTPError",
     "evalue": "HTTP Error 404: Not Found",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mHTTPError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[13], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m getannotation(pdb, fab_list, \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdata/sabdab_download/\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mpdb\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[1;32mc:\\Users\\marce\\FS4_BioInfo_lokal\\group04-team04\\common\\sabdab_downloader.py:148\u001b[0m, in \u001b[0;36mgetannotation\u001b[1;34m(entry, fab_list, out_path)\u001b[0m\n\u001b[0;32m    146\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m Hchain\u001b[38;5;241m!=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNA\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m    147\u001b[0m     out_file \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin( out_path, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m_\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m_VH.ann\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m%\u001b[39m(entry,Hchain) )\n\u001b[1;32m--> 148\u001b[0m     urllib\u001b[38;5;241m.\u001b[39mrequest\u001b[38;5;241m.\u001b[39murlretrieve(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhttps://opig.stats.ox.ac.uk/webapps/abdb/entries/\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m/annotation/\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m_\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m_VH.ann\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m%\u001b[39m(entry,entry,Hchain), out_file)\n\u001b[0;32m    149\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39misfile(out_file):\n\u001b[0;32m    150\u001b[0m         Retrieved \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(out_file)\u001b[38;5;241m.\u001b[39mread()\n",
      "File \u001b[1;32mc:\\Users\\marce\\anaconda3\\Lib\\urllib\\request.py:240\u001b[0m, in \u001b[0;36murlretrieve\u001b[1;34m(url, filename, reporthook, data)\u001b[0m\n\u001b[0;32m    223\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m    224\u001b[0m \u001b[38;5;124;03mRetrieve a URL into a temporary location on disk.\u001b[39;00m\n\u001b[0;32m    225\u001b[0m \n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    236\u001b[0m \u001b[38;5;124;03mdata file as well as the resulting HTTPMessage object.\u001b[39;00m\n\u001b[0;32m    237\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m    238\u001b[0m url_type, path \u001b[38;5;241m=\u001b[39m _splittype(url)\n\u001b[1;32m--> 240\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m contextlib\u001b[38;5;241m.\u001b[39mclosing(urlopen(url, data)) \u001b[38;5;28;01mas\u001b[39;00m fp:\n\u001b[0;32m    241\u001b[0m     headers \u001b[38;5;241m=\u001b[39m fp\u001b[38;5;241m.\u001b[39minfo()\n\u001b[0;32m    243\u001b[0m     \u001b[38;5;66;03m# Just return the local path and the \"headers\" for file://\u001b[39;00m\n\u001b[0;32m    244\u001b[0m     \u001b[38;5;66;03m# URLs. No sense in performing a copy unless requested.\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\marce\\anaconda3\\Lib\\urllib\\request.py:215\u001b[0m, in \u001b[0;36murlopen\u001b[1;34m(url, data, timeout, cafile, capath, cadefault, context)\u001b[0m\n\u001b[0;32m    213\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m    214\u001b[0m     opener \u001b[38;5;241m=\u001b[39m _opener\n\u001b[1;32m--> 215\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m opener\u001b[38;5;241m.\u001b[39mopen(url, data, timeout)\n",
      "File \u001b[1;32mc:\\Users\\marce\\anaconda3\\Lib\\urllib\\request.py:521\u001b[0m, in \u001b[0;36mOpenerDirector.open\u001b[1;34m(self, fullurl, data, timeout)\u001b[0m\n\u001b[0;32m    519\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m processor \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mprocess_response\u001b[38;5;241m.\u001b[39mget(protocol, []):\n\u001b[0;32m    520\u001b[0m     meth \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(processor, meth_name)\n\u001b[1;32m--> 521\u001b[0m     response \u001b[38;5;241m=\u001b[39m meth(req, response)\n\u001b[0;32m    523\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m response\n",
      "File \u001b[1;32mc:\\Users\\marce\\anaconda3\\Lib\\urllib\\request.py:630\u001b[0m, in \u001b[0;36mHTTPErrorProcessor.http_response\u001b[1;34m(self, request, response)\u001b[0m\n\u001b[0;32m    627\u001b[0m \u001b[38;5;66;03m# According to RFC 2616, \"2xx\" code indicates that the client's\u001b[39;00m\n\u001b[0;32m    628\u001b[0m \u001b[38;5;66;03m# request was successfully received, understood, and accepted.\u001b[39;00m\n\u001b[0;32m    629\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;241m200\u001b[39m \u001b[38;5;241m<\u001b[39m\u001b[38;5;241m=\u001b[39m code \u001b[38;5;241m<\u001b[39m \u001b[38;5;241m300\u001b[39m):\n\u001b[1;32m--> 630\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mparent\u001b[38;5;241m.\u001b[39merror(\n\u001b[0;32m    631\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhttp\u001b[39m\u001b[38;5;124m'\u001b[39m, request, response, code, msg, hdrs)\n\u001b[0;32m    633\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m response\n",
      "File \u001b[1;32mc:\\Users\\marce\\anaconda3\\Lib\\urllib\\request.py:559\u001b[0m, in \u001b[0;36mOpenerDirector.error\u001b[1;34m(self, proto, *args)\u001b[0m\n\u001b[0;32m    557\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m http_err:\n\u001b[0;32m    558\u001b[0m     args \u001b[38;5;241m=\u001b[39m (\u001b[38;5;28mdict\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdefault\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhttp_error_default\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;241m+\u001b[39m orig_args\n\u001b[1;32m--> 559\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_call_chain(\u001b[38;5;241m*\u001b[39margs)\n",
      "File \u001b[1;32mc:\\Users\\marce\\anaconda3\\Lib\\urllib\\request.py:492\u001b[0m, in \u001b[0;36mOpenerDirector._call_chain\u001b[1;34m(self, chain, kind, meth_name, *args)\u001b[0m\n\u001b[0;32m    490\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m handler \u001b[38;5;129;01min\u001b[39;00m handlers:\n\u001b[0;32m    491\u001b[0m     func \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(handler, meth_name)\n\u001b[1;32m--> 492\u001b[0m     result \u001b[38;5;241m=\u001b[39m func(\u001b[38;5;241m*\u001b[39margs)\n\u001b[0;32m    493\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m result \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    494\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[1;32mc:\\Users\\marce\\anaconda3\\Lib\\urllib\\request.py:639\u001b[0m, in \u001b[0;36mHTTPDefaultErrorHandler.http_error_default\u001b[1;34m(self, req, fp, code, msg, hdrs)\u001b[0m\n\u001b[0;32m    638\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mhttp_error_default\u001b[39m(\u001b[38;5;28mself\u001b[39m, req, fp, code, msg, hdrs):\n\u001b[1;32m--> 639\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m HTTPError(req\u001b[38;5;241m.\u001b[39mfull_url, code, msg, hdrs, fp)\n",
      "\u001b[1;31mHTTPError\u001b[0m: HTTP Error 404: Not Found"
     ]
    }
   ],
   "source": [
    "getannotation(pdb, fab_list, f\"data/sabdab_download/{pdb}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
